cpp/src/io/json/json_column.cu

```diff
@@ -0,0 +1,810 @@
+  //   unique_copy_by_key {parent_node_id} {row_offset} to
+  //   col[parent_col_id].child_offsets[row_offset[parent_node_id]]
+  rmm::device_uvector<NodeIndexT> original_col_ids(col_ids, stream);  // make a copy
+  thrust::copy(rmm::exec_policy(stream), col_ids.begin(), col_ids.end(), original_col_ids.begin());
+  rmm::device_uvector<size_type> node_ids(row_offsets.size(), stream);
+  thrust::sequence(rmm::exec_policy(stream), node_ids.begin(), node_ids.end());
+  thrust::stable_sort_by_key(
+    rmm::exec_policy(stream), col_ids.begin(), col_ids.end(), node_ids.begin());
+  auto ordered_parent_node_ids =
+    thrust::make_permutation_iterator(tree.parent_node_ids.begin(), node_ids.begin());
+  auto ordered_row_offsets =
+    thrust::make_permutation_iterator(row_offsets.begin(), node_ids.begin());
+  thrust::for_each_n(
+    rmm::exec_policy(stream),
+    thrust::counting_iterator<size_type>(0),
+    col_ids.size(),
+    [num_nodes = col_ids.size(),
+     ordered_parent_node_ids,
+     ordered_row_offsets,
+     original_col_ids = original_col_ids.begin(),
+     col_ids          = col_ids.begin(),
+     row_offsets      = row_offsets.begin(),
+     node_categories  = tree.node_categories.begin(),
+     d_columns_data   = d_columns_data.begin()] __device__(size_type i) {
+      auto parent_node_id = ordered_parent_node_ids[i];
+      if (parent_node_id != parent_node_sentinel and node_categories[parent_node_id] == NC_LIST) {
+        // unique item
+        if (i == 0 ||
+            (col_ids[i - 1] != col_ids[i] or ordered_parent_node_ids[i - 1] != parent_node_id)) {
+          // scatter to list_offset
+          d_columns_data[original_col_ids[parent_node_id]]
+            .child_offsets[row_offsets[parent_node_id]] = ordered_row_offsets[i];
+        }
+        // TODO: verify if this code is right. check with more test cases.
+        if (i == num_nodes - 1 || (col_ids[i] != col_ids[i + 1])) {
+          // last value of list child_offset is its size.
+          d_columns_data[original_col_ids[parent_node_id]]
+            .child_offsets[row_offsets[parent_node_id] + 1] = ordered_row_offsets[i] + 1;
+        }
+      }
+    });
+
+  // 5. scan on offsets.
+  for (auto& [id, col_ref] : columns) {
+    auto& col = col_ref.get();
+    if (col.type == json_col_t::StringColumn) {
+      thrust::inclusive_scan(rmm::exec_policy(stream),
+                             col.string_offsets.begin(),
+                             col.string_offsets.end(),
+                             col.string_offsets.begin(),
+                             thrust::maximum<json_column::row_offset_t>{});
+    } else if (col.type == json_col_t::ListColumn) {
+      thrust::inclusive_scan(rmm::exec_policy(stream),
+                             col.child_offsets.begin(),
+                             col.child_offsets.end(),
+                             col.child_offsets.begin(),
+                             thrust::maximum<json_column::row_offset_t>{});
```

Comment:
Do you mean that, we could hold the data is same variable `offset` instead of two different variables?
Note that, string offsets will be translated, but list child_offsets will moved during cudf column construction (So, we need to use different `mr` for memory allocation of these variables).